### PR TITLE
documentation proposal for cocotb extension modules

### PR DESCRIPTION
**Pull request by Martoni**
This is a guideline proposal to explain how to write an extension protocol module.
Related to [this](https://github.com/cocotb/cocotb/issues/1008) discussion.

I think that a howto is necessary to encourage people to publish their modules.

This PR is mainly to start the discussion.

### CODE REVIEW

File: documentation/source/protocol_extension.rst [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+setup.py is a python configuration file that be used for packaging. An example
+is given above ::
+
+  try:
+    from setuptools import setup
+  except ImportError:
+    from distutils.core import setup
+
+  setup(name='cocotb_spi',
```
Comment (themperek) [inline documentation/source/protocol_extension.rst]: Should the package be called `cocotb-spi`? We have few in such convention.

File: documentation/source/protocol_extension.rst [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+      version='0.1',
+      py_modules=['cocotb_spi'],
+      )
+
+With this config file, python module can be installed with ::
+
+  $ python setup.py install
```
Comment (themperek) [inline documentation/source/protocol_extension.rst]: Another option is `pip install [-e] ` that would be consistent with cocotb documentation.

**Comment by themperek** [comment]
In general, it sound like a good start for me.

File: documentation/source/protocol_extension.rst [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+-----------
+
+A Bus monitor is mainly a passive class that will monitor the bus.
+
+TODO: explain monitor
+
+Python module
+-------------
+
+The module file hierarchy should be as following ::
```
Comment (eric-wieser) [inline documentation/source/protocol_extension.rst]: I think we should link to the pypa distribution docs here, rather than repeating them

File: documentation/source/protocol_extension.rst [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+With python module source code in directory cocotb_spi and module configuration
+in setup.py
+
+setup.py is a python configuration file that be used for packaging. An example
+is given above ::
+
+  try:
+    from setuptools import setup
+  except ImportError:
+    from distutils.core import setup
```
Comment (eric-wieser) [inline documentation/source/protocol_extension.rst]: Does cocotb do this? If not, we may as well assume setuptools

File: documentation/source/protocol_extension.rst [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+  $ python setup.py install
+
+or for development::
+
+  $ python setup.py develop
+
+
+Then module can be imported with import keyword in your testbench ::
+
+  import cocotb_spi
```
Comment (eric-wieser) [inline documentation/source/protocol_extension.rst]: An alternative would be `cocotb.ext.spi`

File: documentation/source/protocol_extension.rst (hunk 1 above)
Comment (Martoni) [reply documentation/source/protocol_extension.rst]: Yes, '-' is better I think. I like the eric proposition with `cocotb.ext.spi` but is it possible with an external package ?

File: documentation/source/protocol_extension.rst (hunk 2 above)
Comment (Martoni) [reply documentation/source/protocol_extension.rst]: I discover the pip option -e, yes it better.

File: documentation/source/protocol_extension.rst (hunk 4 above)
Comment (Martoni) [reply documentation/source/protocol_extension.rst]: As I understand setuptools, the package `distutils.core` is for old python version (python2) and new python version use from `setuptools`.

File: documentation/source/protocol_extension.rst (hunk 5 above)
Comment (Martoni) [reply documentation/source/protocol_extension.rst]: If it's possible with an external package it could be better yes

File: documentation/source/protocol_extension.rst (hunk 4 above)
Comment (cmarqu) [reply documentation/source/protocol_extension.rst]: Make sure the ``setup.py`` contains lines like these two: https://github.com/wallento/cocomod-uart/blob/64f7af8922ea67efda4a1a3c185ce2ecaa45e413/setup.py#L18 and https://github.com/wallento/cocomod-uart/blob/64f7af8922ea67efda4a1a3c185ce2ecaa45e413/setup.py#L26.

File: documentation/source/protocol_extension.rst (hunk 5 above)
Comment (eric-wieser) [reply documentation/source/protocol_extension.rst]: PEP420 makes it easy to make `cocotb.ext` a namespace package, which allows multiple distributions (things with a `setup.py`) to exist within the same package - all that is required is that third-party extensions do _not_ include a `cocotb/__init__.py` or a `cocotb/ext/__init__.py`.

File: documentation/source/protocol_extension.rst (hunk 4 above)
Comment (Martoni) [reply documentation/source/protocol_extension.rst]: I discover the module cocomod-uart. Should not it be referenced in github [cocotb organisation ](https://github.com/cocotb)?

File: documentation/source/protocol_extension.rst (hunk 5 above)
Comment (Martoni) [reply documentation/source/protocol_extension.rst]: have you got a template example for that @eric-wieser ? I'm trying to do that without success [here]( https://github.com/Martoni/cocotb_ext_spi.git).

**Comment by Martoni** [comment]
I improved the document according to discussion. I'm trying it with this package.
I can't manage to setup @eric-wieser proposal : cocotb.ext.spi

File: documentation/source/protocol_extension.rst (hunk 1 above)
Comment (eric-wieser) [reply documentation/source/protocol_extension.rst]: This is an entirely different question - there are two things to decide here:

* The module name, as in `import module_name`. This is what goes in `py_modules`, and the name of the python file
* The distribution name, as in `pip install distribution-name`. This is what goes here in setup.py

@themperek is arguing that we don't need to choose these two names to be the same. While this is true, users might find it easier to realize that `import cocotb_spi` means they need to run `pip install cocotb_spi` instead of `pip install cocotb-spi`, so it might not be worth making the names different.

I'll see if I can get anywhere with an example project using `cocotb.ext.spi`.

File: documentation/source/protocol_extension.rst (hunk 5 above)
Comment (eric-wieser) [reply documentation/source/protocol_extension.rst]: It's looking like that might be a bad idea after all. I'll try to collect some example of other projects choosing to do this here:

* `flask.ext` was removed: https://github.com/pallets/flask/issues/1135#issuecomment-61860862

File: documentation/source/protocol_extension.rst (hunk 4 above)
Comment (cmarqu) [reply documentation/source/protocol_extension.rst]: > I discover the module cocomod-uart. Should not it be referenced in github [cocotb organisation ](https://github.com/cocotb)?

It's already kinda tied in by this new "Used by:" button that GitHub shows in the top row that leads to https://github.com/cocotb/cocotb/network/dependents

File: documentation/source/protocol_extension.rst (hunk 5 above)
Comment (Martoni) [reply documentation/source/protocol_extension.rst]: Ok, glad to hear that. It was really painful to do (without success !)

File: documentation/source/protocol_extension.rst (hunk 1 above)
Comment (Martoni) [reply documentation/source/protocol_extension.rst]: @wallento use `cocomod` as prefix for his [protocol package](https://github.com/wallento/cocomod-uart). Is there other package that using this `cocomod` name or is it only name from @wallento packages ?

File: documentation/source/protocol_extension.rst (hunk 5 above)
Comment (eric-wieser) [reply documentation/source/protocol_extension.rst]: A much simpler option would be `cocotb_ext.spi`, but I'm not sure that's an improvement over `cocotb_spi`

**Comment by Martoni** [comment]
I wrote a minimal module for SPI test with some documentation explaination [here](https://github.com/Martoni/cocomod-spi).
I followed the wallento module [cocomod-uart](https://github.com/wallento/cocomod-uart). It seemed the most correspond to the desired extension model.
All comments and contribution are welcome :)

File: documentation/source/protocol_extension.rst (hunk 1 above)
Comment (wallento) [reply documentation/source/protocol_extension.rst]: And yes, it should use hypen ('-') for the package name

Comment (wallento) [reply documentation/source/protocol_extension.rst]: Ah, sorry, in sphinx its sphinxcontrib, sphinx.ext is in the source tree too

**Comment by eric-wieser** [comment]
IMO the module prefix should contain `cocotb`, which makes it very clear that its for interop with cocotb, unlike `cocomod` which is less clear.

**Comment by Martoni** [comment]
@eric-wieser 
>   IMO the module prefix should contain cocotb, 

Then for uart example : 
- `cocotb-uart` ?
- `cocotb-ext-uart` ?
- `cocotb-mod-uart` ?

Or with dot '.' : 
- `cocotb.uart` ?
- `cocotb.ext.uart` ?
- `cocotb.mod.uart` ?

But it's more a problem with dot to package it correctly through Python2/Python3 no ?

**Comment by wallento** [comment]
The package name can be the first. The second is how the modules are organized and "cocotb.*" doesn't work as it would require the main cocotb package to not use the toplevel `__init__.py`. I don't want to rule it out, but it extremely inconvenient and most probably breaks things.

@eric-wieser understood, how about cocotbmod.* cocotbext.*? The idea of cocomod was that its not the testbench (cocotb) but the modules for the testbench (cocomod). Any other suggestions?

**Comment by eric-wieser** [comment]
Removing `cocotb/__init__.py` sounds like a non-starter to me - I'm happy to outright rule it out.

Happy with either, although would consider an underscore, as `cocotb_mod.mymod`, `cocotb_ext.mymod`, or perhaps to align with Sphinx, `cocotb_contrib.mymod`.

The latter has the advantage of being a coverall that spans both testbench modules (eg protocol drivers), extensions (eg wavedrom-like things), and miscellaneous utilities (eg constrained randomizers)

**Comment by wallento** [comment]
I think `cocotbext` is the best, no underscore. Let's maybe give it a first try for a couple of modules and we can change it later anyhow.

**Comment by Fatsie** [comment]
I don't like the *contrib proposal. As hinted during ORConf I am working on Wishbone support for cocotb. I don't mind become maintainer of that but I would still like it to be under the cocotb umbrella and have the repo hosted under https://github.com/cocotb.

**Comment by eric-wieser** [comment]
@Fatsie: To be clear, are you expressing opposition to the package name being `cocotb_contrib`, or to creating a separate github org like https://github.com/sphinx-contrib

**Comment by Fatsie** [comment]
I prefer `ext` or `mod` and don't like `contrib` in the name.
I propose to host the wishbone repo under https://github.com/cocotb/cocotbext-wishbone

**Comment by wallento** [comment]
okay, so lets go with this naming. I have prepared the infrastructure with the old naming here https://github.com/wallento/cocomod-wishbone based on @Fatsie's repository. We should prepare the infrastructure there and start playing with it.

**Comment by Fatsie** [comment]
Stefan Wallentowitz schreef op do 14-11-2019 om 06:49 [-0800]:
> okay, so lets go with this naming. I have prepared the infrastructure with the old naming here https://github.com/wallento/cocomod-wishbone based on @Fatsie's repository. We should prepare the infrastructure there and start playing with it.

One of the annoying aspects of having separate repos for each module is that each repo will need to setup their own unit test framework and CI/CD...

greets,
Staf.

**Comment by eric-wieser** [comment]
@Fatsie: We could copy `sphinx-contrib` here and create a cookie-cutter project, like https://github.com/sphinx-contrib/cookiecutter

They've pinned it on their org page too, https://github.com/sphinx-contrib, to make it easy to find.

**Comment by imphil** [comment]
So let's try to wrap this up.


**Related work:**
- Flask ([Docs](https://flask.palletsprojects.com/en/1.1.x/extensiondev/)): flask_* as package name, flask-* or *-flask as distribution name.
- Sphinx [doesn't document](https://www.sphinx-doc.org/en/master/extdev/index.html) the naming of external extensions, but de-facto most extensions use `sphinxcontrib.*` as package name, and `sphinxcontrib-*` as distribution ("pip") name. [For in-tree extensions](https://www.sphinx-doc.org/en/master/usage/configuration.html#confval-extensions) Sphinx uses the `sphinx.ext.*`  namespace. The namespacing [is implemented as pkg_resource-style namespace packages](https://github.com/sphinx-contrib/cookiecutter/blob/master/%7B%7Bcookiecutter.project_name%7D%7D/sphinxcontrib/__init__.py#L12) (now deprecated).

**Package name: Use namespaces?**
  - Sphinx uses them and it seems to work for them.
  - Flask did use them, and moved back to non-namespaced packages (https://github.com/pallets/flask/pull/1484).

**Package name: Use subnamespaces? `cocotbXXX.*` or `cocotb.XXX.*`?**
Subnamespaces (`cocotb.XXX.*`) require modifications/removal of the `cocotb/__init__.py` file, so that's not really an option.
  
**Implementing namespaced packages**
All options are outlined in https://packaging.python.org/guides/packaging-namespace-packages/.

  - Native namespace packages, as described in PEP420, are not supported in Python 2.7. They are compatible with pkgutil-style namespaced packages (below), so this scheme can be used for Python3-only packages.
  - ["pkgutil-style namespace packages"](https://packaging.python.org/guides/packaging-namespace-packages/#pkgutil-style-namespace-packages) require an idential `__init__.py` everywhere, but should work in all supported configurations.
  - [pkg_resources-style namespace packages](https://packaging.python.org/guides/packaging-namespace-packages/#pkg-resources-style-namespace-packages) are deprecated.

**Hosting/Discoverability**
- No matter where extensions are hosted, they are discoverable in GitHub https://github.com/cocotb/cocotb/network/dependents and similar sites (e.g. https://libraries.io/pypi/cocotb/dependents)
- Personal opinion: It's not important if an extension is hosted in the cocotb GitHub organization; we can do that for some extensions if extension authors prefer that, but it's not a requirement or even recommended.

**Naming**
Let's come back to that later once we have answers to the namespace or not question.


So let's tackle this step-by-step to get consensus. **First: Should we use namespaces for package names, or not? That is, do you prefer and/or have good reasons for:**

```python3
# with namespaces (sphinx-style)
import cocotbXXX.myextension

# or
# without namespaces (flask-style)
import cocotbXXX_myextension
```

File: documentation/source/protocol_extension.rst (hunk 1 above)
Comment (ktbarrett) [reply documentation/source/protocol_extension.rst]: @eric [you can't extend from the cocotb namespace](https://packaging.python.org/guides/packaging-namespace-packages/#pkgutil-style-namespace-packages). We have actually tried this recently.

**Comment by wallento** [comment]
I prefer namespaces as it is a clean nice solution. Did you see any information about what exactly didn't work with flask? Re implementation I don't mind about 2.7, but pkg-util style is a good thing and it has the advantage that it explicitly codes the information.

So, namespace +1

**Comment by Fatsie** [comment]
I also prefer use of namespace.

**Comment by Martoni** [comment]
Namespace seems to be clear. But if no namespace easier it's ok for me also.

File: documentation/source/protocol_extension.rst [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+
+TODO: give an example.
+
+Bus Monitor
+-----------
+
+A Bus monitor is mainly a passive class that will monitor the bus.
+
+TODO: explain monitor
+
```
Comment (cmarqu) [inline documentation/source/protocol_extension.rst]: I feel here is a good place to add the following information based on https://github.com/cocotb/cocotb/pull/1190#discussion_r347848900

```
Handling Errors
---------------

It may not be clear when to raise an exception and when to use ``log.error()``.

Use ``raise`` if the caller called your function in an invalid way,
and it doesn't make sense to continue.

Use ``log.error()`` if the hardware itself is misbehaving,
and throwing an error immediately would leave it an invalid state.

Even if you do eventually throw an exception (if you weren't able to do it immediately),
you should also ``log.error()`` so that the simulation time of when things went wrong
is recorded.

TL;DR: ``log.error()`` is for humans only, raise is for both humans and code.
```

**Comment by imphil** [comment]
Thanks for the vote. So namespaces it is: `cocotbXXX.myextension`. 

Now step 2: what should XXX be?

Options we've heard so far:
- `cocomod.*`
- `cocotbmod.*`
- `cocotbext.*`
- `cocotbcontrib.*`

Please let me know your preferences!

**Comment by cmarqu** [comment]
I have a slight preference for ``cocotb_ext``.

**Comment by imphil** [comment]
So on the namespace names, we have the following votes so far:

- cocotb_ext.*: @themperek, @cmarqu 
- cocotbext.*: @wallento, @imphil

**Comment by imphil** [comment]
Moving on to the final naming part, package names:

If we go with cocotb_ext as namespace, we could have:

- cocotb-ext-YOURPACKAGE
- cocotb_ext-YOURPACKAGE

If we go with cocotbext as namespace, we would have:
- cocotbext-YOURPACKAGE


Any opinions on that? This should be the last vote before we are done with naming.

**Comment by Fatsie** [comment]
I prefer cocotbext.*

**Comment by eric-wieser** [comment]
> If we go with cocotb_ext as namespace, we could have

`cocotb_ext.YOURPACKAGE` is a totally legal name too - see https://pypi.org/project/backports.functools-lru-cache/

**Comment by imphil** [comment]
@eric-wieser what do you mean by "github package name"? The name of the repository, including the question of where the code lives? That would be a question for another day, taking one step at a time here.

**Comment by eric-wieser** [comment]
@imphil: exactly. Edited my comment, feel free to edit it more as you see fit to focus discussion

**Comment by Martoni** [comment]
I prefer cocotbext.* too

**Comment by ktbarrett** [comment]
I like either `cocotb_ext` or `cocotbext`, preference is for the former.

File: documentation/source/protocol_extension.rst (hunk 6 above)
Comment (Martoni) [reply documentation/source/protocol_extension.rst]: Added

**Comment by Martoni** [comment]
I updated PR and my spi module for following naming convention:
package name:

>  cocotbext-spi

import command :  
```
import cocotbext.spi
```
The cocotbext-spi module can be [seen here](https://github.com/Martoni/cocotbext-spi).

**Comment by imphil** [comment]
Thanks for the discussion everyone, and thanks for starting this @Martoni! 

The cocotb maintainers have discussed this today and we have finalized the decision on naming:

- "cocotbext" for the namespace
- "cocotbext-" to prefix the distribution (package) name.

This PR already has the right naming, so that part is good to go.

When I looked through the text I saw a couple grammar/wording nits I'd like to clean up, it's probably easier to do that right away and not comment on each small thing individually. I'll do that tonight and update this PR with my changes.

**Comment by imphil** [comment]
I did reword the contents in this PR and opened https://github.com/cocotb/cocotb/pull/1344 with the updated wording. I thought it would be clearer to do that in a separate PR and not overwrite the changes done in this PR. @Martoni can you pleae have a look there?
